---
title: RNeXML: Parsing and Serializing the Next Generation of Phyloinformatic Data in R
author: 
  - name: Carl Boettiger
    affiliation: cstar
    email: cboettig@gmail.com
    footnote: Corresponding author
  - name: Hilmar Lapp
    affiliation: NESCent
  - name: Scott Chamberlain
    affiliation: SimonFraser
  - name: Rutger Vos
    affiliation: NCB
address: 
  - code: cstar
    address: | 
      Center for Stock Assessment Research, 
      Department of Applied Math and Statistics, 
      University of California, Mail Stop SOE-2,
      Santa Cruz, CA 95064, USA
  - code: NESCent
    address: | 
      National Evolutionary Synthesis Center, Durham, NC, USA
  - code: SimonFraser
    address: |
      Department of Biology, Simon Fraser University, Burnaby, Canada
  - code: NCB
    address: | 
      NCB Naturalis, Leiden, the Netherlands
abstract: | 
      1. NeXML promises to be the next generation of phylogenetic
      informatics data and metadata exchange format, offering substantial
      improvements in reliability, extensibility, and richness over
      earlier standards such as Nexus.

      2. In this paper we present the RNeXML package which provides
      a bridge between this rich and reliable data format and the
      extensive library of phylogenetic tools already available in R,
      with a particular emphasis on comparative methods. 


      3. RNeXML allows R users to both read and write NeXML files. Yet
      RNeXML is much richer than a basic parser: here we illustrate
      how the package facilitates adding and manipulating semantic 
      annotations.

      4. An essential feature of NeXML is to provide both the flexibility
      to be continually extended to meet the emerging needs of the phylogenetics
      community while remaining backwards-compatible with software developed
      for earlier versions.  Here we illustrate how RNeXML makes it easy
      to extend NeXML in this way.  
      
      5. RNeXML is perfectly suited for users to preserve and maintain their
      personal archives of phylogenetic and comparative trait data in consistent,
      query-able, and cross-platform format, or leverage the rapidly growing 
      public community archives of rich NeXML files.  

---



Though R provides the most extensive and widely used collection of comparative phylogenetic methods tools, R users lack a convenient, extensible format to exchange and archive comparative methods data.  


## Background

### Challenges with existing formats
See [Vos _et al._ 2012].  

### Challenges with R formats

- Most R users rely on the nexus inherits the limitations of nexus file format

See [phylobase vignette, 2010]




### NeXML

See [Vos _et al._ 2012].

Mention of data sharing challenges and opportunities.  See [Drew et al 2013](10.1371/journal.pbio.1001636)




Why RNeXML?

* Why NeXML -- because Nexus standard isn't a standard.  Developers extend it arbitrarily to meet their needs, making it incompatible.  Cannot represent horizontal gene transfer, ambiguous use of symbols, cannot be extended reliably, no validator.  NeXML addresses all of these problems [Vos _et al._ 2012] with a state-of-the-art data management format.  

* Why RNeXML: Huge ecosystem of phylogenetics packages. (1) Read nexml.  All would benefit by being able to read NeXML, both as a robust standard free of the pitfalls of Nexus, and more importantly, as a much richer source of data thanks to its extendable annotation model.  (2) Write nexml.  R packages and R users add rich information to the phylogenies they work with, but need a platform independent data exchange format to share and archive this. (3) Annotate and extend nexml.  

* as your own personal treebase

* Adding metadata to capture workflow





## Introduction to RNeXML


A nexml file can be thought of as having three components: (1) phylogeny data (2) character-trait data, and (3) metadata. `RNeXML` seeks to provide a seemless way to convert these to and from their corresponding R object representations, such as `ape::phylo` phylogenies, character matrices (`data.frames`) and metadata lists (or for advanced users, `XMLInternalDocument` and `rrdf` triplestores). Metadata can provide information about the study in which the phylogeny was generated or applied (such as authors, citations to publications, timestamps, and even descriptions of methods), but can also be used to annotate individual data components.  We illustrate how metadata annotations can be added to the operational taxonomic units of the phylogeny and can be used to extend the data representation to new uses such as stochastic character maps.  

The current stable version of the `RNeXML` package can be installed from the CRAN library <!-- Actually not yet -->
using the standard 

```r
install.packages("RNeXML")
```

The latest stable development version can be installed from Github using the `devtools` package:

```r
library(devtools)
install_github("RNeXML", "ropensci")
```


## Working with phylogenetic trees

#### Read methods

```{r echo=FALSE}
library(RNeXML)
```

We begin by reading in an example NeXML file provided with the package.  

```{r}
f <- system.file("examples", "trees.xml", package="RNeXML")
nex <- nexml_read(f)
```

(For convenience `nexml_read` is also aliased as `read.nexml`, for
users familiar with the somewhat analogous `read.nexus` function).
The resulting `nex` object is an R object representation of the NeXML.
From this object we can extract any phylogenies it contains in the
`ape::phylo` format:

```{r}
phy <- get_trees(nex)
```

We can then leverage the rich suite of methods available from the R
phylogenetics community (including coercion into other representations,
such as `ouch` [@ouch] and `phylobase` [@phylobase] phylogeny
formats). For instance, here we use the standard plotting function from
the `ape` package to plot the resulting `ape::phylo` object returned by
`get_trees`.



``{r}
plot(phy)
```

NeXML can not only contain multiple phylogenetic trees in a single file,
but can also group sets of trees into separate blocks.  The `get_trees`
function attempts to return the most sensible R object for NeXML: if
only a single tree is available, it will return an `ape::phylo` object;
if multiple trees are available in a single `trees` block, it will return
an `ape::multiPhylo` object (a list of `phylo` objects).  If multiple
blocks are available, it will return a list of `multiPhylo` blocks.
While most users will find this behavior convenient, there are times
(such as looping over many NeXML files) when having a variable return
type is undisirable.  To always get back a list of `multiPhylo` objects,
even when only one tree is available, use `get_trees_list()`.

<!-- have we already said all this above?--> The ability to read in
phylogenetic trees in the `nexml` format opens up a wide and rapidly
growing array of resources to the R user.  For instance, a user can access
all the phylogenies available in TreeBASE through the `nexml` format.
`RNeXML` can read directly from a URL:

```{r eval=FALSE}
nex = nexml_read("https://raw.github.com/rvosa/supertreebase/master/data/treebase/S100.xml")
```

Previously, this was possible using the `treebase` package [@treebase]
for R, which relied on the nexus parser provided in the `ape` package.
Due to inconsistencies present in the Nexus standard [Vos _et al._
2012], a significant number <!-- State explicitly --> of TreeBASE nexus
files could not be read in this way.  Using the more precisely defined
NEXML standard, every valid NeXML file on TreeBASE can now successfully
be parsed.



#### Character data 

Like nexus files, NeXML files can also contain character data.  While
most researchers will be familiar with the use of character matrices in
nexus files as a container for molecular sequence data used to infer
the phylogenetic tree, we find this ideal for character data used in
comparative methods popular in R. Methods to get extract character data
work much like those for the phylogenetic data:

```{r }
nex <- nexml_read("comp_analysis.xml")
get_characters(nex)
```

Returns a `data.frame` with columns as characters and rows as
taxa. Continuous and discrete characters are represented as seperate
blocks in NeXML files, but will be combined as seperate columns of a
single data frame by `get_characters()` if they correspond to the same
taxa. The `get_characters_list()` function will consistently recieve
a list of data frames corresponding to each character matrix block of
the NeXML.


#### Write methods 

Generating NeXML files from R is likewise straight forward.  Here we
write a phylogeny from the `ape::phylo` format out to a NeXML file:

```{r}
data(bird.orders)
nexml_write(bird.orders, file = "birds.xml")
```

The first argument to `nexml_write` can be a `phylo` phylogeny,
`multiPhylo` list of phylogenies, or a `nexml` object -- anything that
can be coerced into the `nexml` class.  As we have seen, NeXML files
can contain multiple phylogenetic trees, as well as character matrices.
We can add both trees  and character data using `nexml_write`,

```{r}
data(geospiza)
nexml_write(trees = geospiza$phy, characters = geospiza$dat, file="geospiza.xml")
```


NeXML files can also be built up bit by bit using the `add_characters`,
`add_trees`, and `add_meta`.  Each of these functions can add data to an
existing `nexml` object or create a new nexml object if none is provided.
Here we first create a `nexml` object containing the phylogeny data,
and then add the character data to it:

```{r}
nex <- add_trees(geospiza$phy)
nex <- add_characters(geospiza$dat, nex)
```

The data need not share the same taxa.  We can append additional
phylogenies and characters corresponding to different taxa:

```{r}
data(primates)
nex <- add_trees(primates$phy, nex)
nex <- add_characters(primates$dat, nex)
```




### Metadata

At the heart of the NeXML standard is the ability to encode arbitrary
metadata in precisely specified, machine-readable formats that maintain
backwards compabitibilty with any existing NeXML parser. NeXML uses the
"Resource Description Framework in Annotations", or RDFa [W3C 2013],
as described in [Vos et al 2011].  These annotations can be used to add 
information to any particular element of the data, (a tree, node, edge,
or taxonomic unit; a character matrix or specific trait, or even another 
metadata annotation or the NeXML data file as a whole). `RNeXML` provides
a range of tools for accessing metadata embedded in a NeXML file.  


### Reading NeXML metadata

To illustrate metadata access, we begin by reading in a nexml file.  


```{r } 
S100 <- nexml_read("https://raw.github.com/rvosa/supertreebase/master/data/treebase/S100.xml")
```

A call to the nexml object prints some metadata summarizing the data structure:

```{r }
S100
```

We can extract all metadata pertaining to the NeXML document as a whole
(annotations of the XML root node, `<nexml>`) with the command

```{r}
meta <- get_metadata(S100) 
```

This returns a named list of available metadata. We can see the kinds
of metadata recorded from the names:

```{r}
names(meta)
```

and can ask for a particular element using the standard list subsetting
mechanism (i.e. either the name of an element or it's numeric position),

```{r}
meta[["dc:title"]]
meta[["dc:publisher"]]
```

All metadata terms must belong to an explicit *namespace* or vocabulary that allows a computer to interpret the term precisely. The prefix (before the `:`) indicates to which vocabulary the term belongs, e.g. `dc` in this case.  The `get_namespaces` function tells us the definition of the vocabulary using a link:

```{r}
prefixes <- get_namespaces(S100)
prefixes["dc"]
```

Following the link we can see that `dc` refers to the "Dublin Core"
vocabulary, and the term `title` refers to the title of a creative work
(as opposed to the title of a person, a legal title, the verb title, etc).
Such explicit definitions allow a computer to reason about these objects,
as we shall see later.  


Other metadata is available 

```{r}
get_taxa(S100)
```


- article citation information

- Summary data: taxa, number of taxa, number of trees, number of tips.  

- Metadata of nodes, complete RDF extraction



This allows us to include multiple trees and character matrices to a single NeXML file.  

<!-- explain OTU blocks somewhere -->

### Taxonomic identifiers

Add taxonomic identifier metadata to the OTU elements:


```{r}
nex <- add_trees(bird.orders)
nex <- nexml_taxize(nex)
```


Replicate examples from [treebase] metadata section: publication information, tree size, taxanomic coverage



### Writing NeXML metadata

Researchers cannot take advantage of NeXML metadata manipulation if
there is no metadata there.  The burden of adding good metadata remains a
significant barrier to increased sharing and reuse of scientific research
[Tenopir et al. 2011].  `RNeXML` seeks to reduce the burden of adding
high quality, machine-readable metadata to NeXML files. This occurs on
several levels:

- Automatic annotation.  

`RNeXML` automatically adds basic metadata by default such as a timestamp
of when the tree was produced and an open data license (CC0, compatible
with deposition on Dryad data repository), which can be configured
as needed.



- Utilities to add common metadata

  - Add a citation to any papers associated with the phylogeny data.  
  - Writing in methodological information on how tree has been generated or analyzed (including R code)
  - Add annotations to any taxon.  

Writing annotations on otus.  Annotations on nodes.  

Temporal, spatial, taxonomic coverage for dryad/knb?


### Extending the NeXML standard through metadata annotation.

No data standard can anticipate all the uses and extensions researchers may one day need. Researchers have repeatedly adapted the existing NEXUS file format to accomidate encoding new kinds of information, but in the process break compatibility between different flavors of NEXUS files, with few reliable methods to distinguish between these flavors.  NeXML provides a solution to this problem through the use of metadata annotation. 

Here we illustrate this process using the example of stochastic character mapping [Huelsenbeck et al 2003]. A stochastic character map is simply an annotation of the branches on a phylogeny, assigning each section of each branch to a particular "state" (typically of a morphological characteristic).  

[Bollback 2006] provides a widely used stand-alone software implementation of this method in the software `simmap`, which modified the standard Newick tree format to express this additional information. This can break compatibility with other software[^1], and creates a format that cannot be interpreted without additional information describing this convention.  By contrast, the NeXML extension is not only backwards compatible but contains a precise and machine-readable description of what it is encoding.  



[^1]: By using the commenting mechanism of the Newick format, it is possible that other software that doesn't also use the comment mechanism for some other such purpose would be able to successfully parse the tree.  However there is no way to guarentee that this is the case or for the data format to describe its use.  

[Revell 2011] describes the `phytools` package for R, which includes utilities for reading, manipulating, and writing `simmap` files.  






NeXML annotation of an edge that switches from state `s2` to state `s1`: 

```xml
<edge about="#e34" id="e34" source="n16" target="n35" length="0.205223">
        <meta xsi:type="nex:LiteralMeta" id="m262" property="simmap:reconstructions" datatype="xsd:string">
          <meta xsi:type="nex:LiteralMeta" id="m260" property="simmap:reconstruction" datatype="xsd:string">
            <meta xsi:type="nex:LiteralMeta" id="m261" property="simmap:char" datatype="xsd:string" content="cr1"/>
            <meta xsi:type="nex:LiteralMeta" id="m252" property="simmap:stateChange" datatype="xsd:string">
              <meta xsi:type="nex:LiteralMeta" id="m253" property="simmap:order" datatype="xsd:decimal" content="1"/>
              <meta xsi:type="nex:LiteralMeta" id="m254" property="simmap:length" datatype="xsd:decimal" content="0.202947550453246"/>
              <meta xsi:type="nex:LiteralMeta" id="m255" property="simmap:state" datatype="xsd:string" content="s2"/>
            </meta>
            <meta xsi:type="nex:LiteralMeta" id="m256" property="simmap:stateChange" datatype="xsd:string">
              <meta xsi:type="nex:LiteralMeta" id="m257" property="simmap:order" datatype="xsd:decimal" content="2"/>
              <meta xsi:type="nex:LiteralMeta" id="m258" property="simmap:length" datatype="xsd:decimal" content="0.00227544954675424"/>
              <meta xsi:type="nex:LiteralMeta" id="m259" property="simmap:state" datatype="xsd:string" content="s1"/>
            </meta>
          </meta>
        </meta>
      </edge>
```

[Revell 2011]: http://doi.org/10.1111/j.2041-210X.2011.00169.x
[Huelsenbeck et al 2003]: http://doi.org/10.1080/10635150390192780 "Stochastic Mapping of Morphological Characters"
[Bollback 2006]: http://doi.org/10.1186/1471-2105-7-88 "SIMMAP: stochastic character mapping of discrete traits on phylogenies."





### Subsetting nexml objects directly

The `RNeXML` package provides many convenient functions to add and extract
information from `nexml` objects in the R environment without requiring
the reader to understand the details of the NeXML data structure and
making it less likely that a user will generate invalid NeXML syntax
that could not be read by other parsers. The `nexml` object we have been using
in all of the examples is built on R's S4 mechanism. Advanced users may
sometimes prefer to interact with the data structure more directly using 
R's S4 class mechanism and subsetting methods. Many R users are more familiar
with the S3 class mechanism (such as in the `ape` package phylo objects)
rather than the S4 class mechanism used in phylogenetics packages such as 
`ouch` and `phylobase`. The `phylobase` vignette provides an excellent introduction
to these data structures.  Users already familiar with subsetting lists and other
S3 objects in R are likely familar with the use of the `$` operator, such as
`phy$edge`. S4 objects simply use an `@` operator instead (but cannot be subset
using numeric arguments such as `phy[[1]]` or named arguments such as phy[["edge"]]).  


The `nexml` object is an S4 object, as are all of its components (slots).  It's 
heirachical structure corresponds exactly with the XML tree of a NeXML file, with 
the single exception that both XML attributes and children are represented as slots.  
S4 objects have constructor functions to initialize them.  We create a new `nexml` 
object with the command:

```{r}
nex <- new("nexml")
```

We can see a list of slots contained in this object with

```{r}
slotNames(nex)
```

Some of these slots have already been populated for us, for instance, the schema version and default namespaces:

```{r}
nex@version
nex@namespaces
```

Recognize that `nex@namespaces` serves the same role as `get_namespaces` function, but provides direct access to the slot data.  For instance, with this syntax we could also overwrite the existing namespaces with `nex@namespaces <- NULL`.  Changing the namespace in this way is not advised.

Some slots can contain multiple elements of the same type, such as `trees`, `characters`, and `otus`.  For instance, we see that 

```{r}
class(nex@characters)
```

is an object of class `ListOfcharacters`, and is currently empty,

```{r}
length(nex@characters)
```

In order to assign an object to a slot, it must match the class definition
of the slot.  We can create a new element of any given class with the
`new` function,

```{r}
nex@characters <- new("ListOfcharacters", list(new("characters")))
```

and now we have a length-1 list of character matrices,

```{r}
length(nex@characters)
```

and we access the first character matrix using the list notation, `[[1]]`. Here we check the class is a `characters` object.  

```{r}
class(nex@characters[[1]])
```

Direct subsetting has two primary use cases: (a) useful in looking up (and possibly editing) a specific value of an element, or (b) when adding metadata annotations to specific elements. Consider the example file


```{r}
f <- system.file("examples", "trees.xml", package="RNeXML")
nex <- nexml_read(f)
```

We can look up the species label of the first `otu` in the first `otus` block:

```{r}
nex@otus[[1]]@otu[[1]]@label
```

We can add metadata to this particular OTU using this subsetting format

```{r}
nex@otus[[1]]@otu[[1]]@meta <- 
  c(meta("skos:note", 
          "This species was incorrectly identified"),
         nex@otus[[1]]@otu[[1]]@meta)
```

Here we use the `c` operator to append this element to any existing meta annotations to this otu.  


### Implementation and architecture

The `RNeXML` package is built for the NeXML Schema, [version 2009],
using the R software environment and the XML package for parsing and
generating XML [Temple Lang 2013, XML], and provides conversions into
the popular `phylo` format provided by the `ape` package [Paradis 2004].
`RNeXML` can automatically check taxonomic names and add taxonomic
identifiers when generating NeXML using methods from the `taxize` package
([Chamberlain 2013]).

<!-- Possibly cite all dependencies and suggests from the DESCRIPTION --> 
<!-- Reference any installation and configuration instructions necessary (e.g. API credentials for figshare) -->


### Quality control

The `RNeXML` package performs severak levels of validation on NeXML files
written from or read into R: validating the XML, and validating against
the current NeXML schema, and performing additional checks through the 
NeXML online validation tool for elements that cannot be enforced by
schema validation alone.  Validation serves to improve consistency in
data encoding which facilitates the use of NeXML across a wide variety
of platforms and software.  A lack of a consistent standard and validation
mechanism has continually plagued the current NEXUS standard and made 
interoperability difficult [Vos et al. 2011].  

Several mechanisms are in place to evaluate the quality of the `RNeXML`
package functions themselves.  In addition to the examples from the
documentation and the automated package checks provided for the R system,
the `RNeXML` package includes a complete test suite using the `testthat`
package [Wickham 2011]. Package development on Github includes continuous 
integration through the `travis` platform, in which the full test suite
is run every time the code is changed.  This not only gives users confidence
that the development copy can be successfully installed, but should assist
other developers seeking to contribute to the package to test that their proposed 
changes have not negatively impacted any of the existing code base.  


### Software reuse and support

The `RNeXML` package is available through the CRAN repository under
a CC0 license. At the time of writing, the current version is `r packageDescription("RNeXML", fields="Version")`. 
The current development
branch can be found on Github at [https://github.com/ropensci/RNeXML].
See the NEWS file for changes in recent versions. Users are encouraged
to submit bug reports or feature requests in the issues log.  Contact
ropensci-discuss@googlegroups.com or the package maintainer for help.

There are many promising directions for further development of the
`RNeXML` package.  The project's Milestones page provides a list of
steps currently planned or under consideration.  In particular, these
milestones include increased support for interactive mode of entering data
(wizards), and support for integration of semantics defined through the
Web Ontology language (OWL), as discussed in the Challenges section.



### Acknowledgements

This project was supported in part by a grant from the Alfred P Sloan
Foundation (CB and SC), NSF Grant DBI-1306697 (CB), ... `RNeXML` is part
of the rOpenSci project, [http://ropensci.org].



[Temple Lang 2013, XML]: http://cran.r-project.org/web/packages/XML/
[Wickham 2011]: http://vita.had.co.nz/papers/testthat.html
[Chamberlain 2013]: http://dx.doi.org/10.12688/f1000research.2-191.v2
[@ouch]: http://cran.r-project.org/web/packages/ouch/
[@phylobase]: http://cran.r-project.org/web/packages/phylobase
[W3C 2013]: http://www.w3.org/TR/xhtml-rdfa-primer/ "RDFa 1.1 Primer - Second Edition: Rich Structured Data Markup for Web Documents"
[Tenopir et al. 2011]: http://doi.org/10.1371/journal.pone.0021101 "Data Sharing by Scientists: Practices and Perceptions"



phylobase::phylo4 was an attempt to solve the problems associated with
the popular ape::phylo format (unpredictable errors, loose and multiple
conflicting specifications).



